File: SourceCode/Java_Generic/ChalmersGU/API/MapSet.java

import java.util.Iterator;
    
// This is how Java does it, in the method Collections.newSetFromMap:
// https://stackoverflow.com/questions/13524608/how-to-make-a-set-backed-by-a-map/13524864

/* *** ODSATag: MapSet *** */
/* *** ODSATag: Header *** */
class MapSet<E> implements Set<E> {
    Map<E, Boolean> internalMap;
    // One alternative possibility would be to use a Map<E, Void>,
    // but then Map.add() and Map.remove() would always return null
    // and it would be impossible to know if the key already existed or not.

    MapSet() {
        // Use a separate chaining hash map as the default:
        internalMap = new SeparateChainingHashMap<>();
    }

    MapSet(Map<E, Boolean> initialMap) {
        if (!initialMap.isEmpty()) throw new IllegalArgumentException("The initial map must be empty!");
        internalMap = initialMap;
    }
/* *** ODSAendTag: Header *** */

/* *** ODSATag: Add *** */
    public boolean add(E x) {
        return internalMap.put(x, true) == null;
    }
/* *** ODSAendTag: Add *** */

/* *** ODSATag: Remove *** */
    public boolean remove(E x) {
        return internalMap.remove(x) != null;
    }
/* *** ODSAendTag: Remove *** */

/* *** ODSATag: Contains *** */
    public boolean contains(E x) {
        return internalMap.containsKey(x);
    }
/* *** ODSAendTag: Contains *** */

    public boolean isEmpty() {
        return internalMap.isEmpty();
    }

    public int size() {
        return internalMap.size();
    }

    public Iterator<E> iterator() {
        return internalMap.iterator();
    }
/* *** ODSAendTag: MapSet *** */


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printSet() {
        System.out.print(internalMap.size() + " { ");
        for (E elem : internalMap) System.out.print(elem + ", ");
        System.out.println("}");
    }

    public static void main(String[] args) {
        MapSet<String> set = new MapSet<>();
        System.out.println("Adding values");
        for (int i=0; i<40; i++) {
            set.add(String.valueOf((char)((i%25)+65)));
            if (i % 5 == 0) set._printSet();
        }
        set._printSet();
        System.out.println("Removing values");
        int i = 7;
        while (!set.isEmpty()) {
            set.remove(String.valueOf((char)((i%25)+65)));
            i += 11;
            if (set.size() % 5 == 0) set._printSet();
        }
    }
/* *** ODSATag: MapSet *** */
}
/* *** ODSAendTag: MapSet *** */

File: SourceCode/Python/ChalmersGU/API/MapSet.py

from API import Map, Set
from SeparateChainingHashMap import SeparateChainingHashMap

#/* *** ODSATag: MapSet *** */
#/* *** ODSATag: Header *** */
class MapSet(Set):
    _internalMap : Map
    # We're using a map from elements to the empty tuple ()

    def __init__(self, initialMap=None):
        if initialMap is None:
            # Use a separate chaining hash map as the default:
            self._internalMap = SeparateChainingHashMap()
        else:
            if not initialMap.isEmpty(): raise ValueError("The initial map must be empty!")
            self._internalMap = initialMap
#/* *** ODSAendTag: Header *** */

#/* *** ODSATag: Add *** */
    def add(self, x) -> bool:
        return self._internalMap.put(x, ()) is None
#/* *** ODSAendTag: Add *** */

#/* *** ODSATag: Remove *** */
    def remove(self, x) -> bool:
        return self._internalMap.remove(x) is not None
#/* *** ODSAendTag: Remove *** */

#/* *** ODSATag: Contains *** */
    def contains(self, x) -> bool:
        return self._internalMap.containsKey(x)
#/* *** ODSAendTag: Contains *** */

    def isEmpty(self) -> bool:
        return self._internalMap.isEmpty()

    def size(self) -> int:
        return self._internalMap.size()

    def __iter__(self):
        return iter(self._internalMap)
#/* *** ODSAendTag: MapSet *** */


######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed

    def _printSet(self):
        print(str(self._internalMap.size()) + " { " +
              ", ".join(str(elem) for elem in self._internalMap) + " }")


if __name__ == '__main__':
    theset = MapSet()
    print("Adding values")
    for i in range(40):
        theset.add(chr((i%25)+65))
        if i % 5 == 0: theset._printSet()
    theset._printSet()
    print("Removing values")
    i = 7
    while not theset.isEmpty():
        theset.remove(chr((i%25)+65))
        i += 11
        if theset.size() % 5 == 0: theset._printSet()
